File: svc-data/src/main/java/com/beancounter/marketdata/markets/MarketService.java
package com.beancounter.marketdata.markets;

import com.beancounter.common.contracts.MarketResponse;
import com.beancounter.common.exception.BusinessException;
import com.beancounter.common.model.Market;
import com.beancounter.marketdata.config.StaticConfig;
import java.util.HashMap;
import java.util.Map;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

/**
 * Verification of Market related functions.
 *
 * @author mikeh
 * @since 2019-03-19
 */
@Service
@Slf4j
public class MarketService implements com.beancounter.client.MarketService {

  private Map<String, Market> markets;
  private final Map<String, String> aliases = new HashMap<>();

  /**
   * Return the Exchange code to use for the supplied input.
   *
   * @param input code that *might* have an alias.
   * @return the alias or input if no exception is defined.
   */
  public String resolveAlias(String input) {
    String alias = aliases.get(input.toUpperCase());
    if (alias == null) {
      return input;
    } else {
      return alias;
    }

  }

  /**
   * Resolves a market via its code property.
   *
   * @param marketCode non-null market code - can also be an alias
   * @return resolved market
   */

  public Market getMarket(String marketCode) {
    return getMarket(marketCode, true);
  }

  public Market getMarket(String marketCode, boolean orByAlias) {
    if (marketCode == null) {
      throw new BusinessException("Null Market Code");
    }
    Market market = markets.get(marketCode.toUpperCase());
    String errorMessage = String.format("Unable to resolve market code %s", marketCode);
    if (market == null && orByAlias) {
      String byAlias = resolveAlias(marketCode);
      if (byAlias == null) {
        throw new BusinessException(errorMessage);
      }
      market = markets.get(byAlias);
    }
    if (market == null) {
      throw new BusinessException(errorMessage);
    }

    return market;
  }

  public MarketResponse getMarkets() {

    return MarketResponse.builder().data(markets.values()).build();
  }

  @Autowired
  public void setMarkets(StaticConfig staticConfig) {
    this.markets = staticConfig.getMarketData();
    for (String marketCode : markets.keySet()) {
      Market market = markets.get(marketCode);
      if (!market.getAliases().isEmpty()) {
        for (String provider : market.getAliases().keySet()) {
          this.aliases
              .put(market.getAliases().get(provider).toUpperCase(), marketCode.toUpperCase());
        }
      }
    }

  }
}

File: svc-data/src/main/kotlin/com/beancounter/marketdata/markets/MarketService.kt
package com.beancounter.marketdata.markets

import com.beancounter.common.contracts.MarketResponse
import com.beancounter.common.exception.BusinessException
import com.beancounter.common.model.Market
import jakarta.transaction.Transactional
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Import
import org.springframework.stereotype.Service
import java.util.Locale

/**
 * Obtain Markets.
 *
 * @author mikeh
 * @since 2019-03-19
 */
@Service
@Import(MarketConfig::class)
@Transactional
class MarketService
    @Autowired
    constructor(
        val marketConfig: MarketConfig
    ) : com.beancounter.client.MarketService {
        private val aliases: MutableMap<String, String> = mutableMapOf()
        private var marketMap: Map<String, Market> = mutableMapOf()

        fun getMarketMap(): Map<String, Market> {
            if (marketMap.isEmpty()) {
                marketMap = marketConfig.getProviders()
                for (marketCode in marketMap.keys) {
                    val market = marketMap[marketCode]
                    setAlias(
                        market,
                        marketCode
                    )
                }
            }
            return marketMap
        }

        private fun setAlias(
            market: Market?,
            marketCode: String
        ) {
            if (market != null && market.aliases.isNotEmpty()) {
                for (provider in market.aliases.keys) {
                    aliases[market.aliases.getValue(provider).uppercase(Locale.getDefault())] =
                        marketCode.uppercase(Locale.getDefault())
                }
            }
        }

        /**
         * Return the Exchange code to use for the supplied input.
         *
         * @param input code that *might* have an alias.
         * @return the alias or input if no exception is defined.
         */
        fun resolveAlias(input: String): String {
            val alias = aliases[input.uppercase(Locale.getDefault())]
            return alias ?: input
        }

        /**
         * Resolves a market via its code property.
         *
         * @param marketCode non-null market code - can also be an alias
         * @return resolved market
         */
        override fun getMarket(marketCode: String): Market =
            getMarket(
                marketCode,
                true
            )

        fun getMarket(
            marketCode: String?,
            orByAlias: Boolean
        ): Market {
            if (marketCode == null) {
                throw BusinessException("Null Market Code")
            }
            var market = getMarketMap()[marketCode.uppercase(Locale.getDefault())]
            val errorMessage =
                String.format(
                    Locale.US,
                    "Unable to resolve market code %s",
                    marketCode
                )
            if (market == null && orByAlias) {
                val byAlias = resolveAlias(marketCode)
                market = marketMap[byAlias]
            }
            if (market == null) {
                throw BusinessException(errorMessage)
            }
            return market
        }

        override fun getMarkets(): MarketResponse = MarketResponse(getMarketMap().values)

        fun canPersist(market: Market): Boolean {
            // Don't persist Mock market assets
            return market.code != "MOCK"
        }
    }